vllm/v1/spec_decode/metrics.py

```diff
@@ -0,0 +1,89 @@
+
+from dataclasses import dataclass
+
+import numpy as np
+
+from vllm.config import SpeculativeConfig
+from vllm.logger import init_logger
+
+logger = init_logger(__name__)
+
+
+@dataclass
+class SpecDecodingStats:
+    num_draft_tokens: int = 0
+    num_accepted_tokens: int = 0
+    num_emitted_tokens: int = 0
+
+    def take(self):
+        copied = SpecDecodingStats(self.num_draft_tokens,
+                                   self.num_accepted_tokens,
+                                   self.num_emitted_tokens)
+        self.reset()
+        return copied
+
+    def reset(self):
+        self.num_draft_tokens = 0
+        self.num_accepted_tokens = 0
+        self.num_emitted_tokens = 0
+
+
+class SpecDecodingMetrics:
+
+    def __init__(self, speculative_config: SpeculativeConfig):
+        self.num_spec_tokens = speculative_config.num_speculative_tokens
+        self.reset()
+
+    def reset(self):
+        self.num_draft_tokens: list[int] = []
+        self.num_accepted_tokens: list[int] = []
+        self.num_emitted_tokens: list[int] = []
+
+    def observe(self, spec_decoding_stats: SpecDecodingStats):
+        self.num_draft_tokens.append(spec_decoding_stats.num_draft_tokens)
+        self.num_accepted_tokens.append(
+            spec_decoding_stats.num_accepted_tokens)
+        self.num_emitted_tokens.append(spec_decoding_stats.num_emitted_tokens)
+
+    def log(self):
+        num_draft_tokens = np.sum(self.num_draft_tokens)
+        num_accepted_tokens = np.sum(self.num_accepted_tokens)
+        num_emitted_tokens = np.sum(self.num_emitted_tokens)
+        max_num_emitted_tokens = get_max_num_emitted_tokens(
+            draft_tokens=num_draft_tokens, k=self.num_spec_tokens)
+        draft_acceptance_rate = (num_accepted_tokens / num_draft_tokens
+                                 if num_draft_tokens > 0 else float("nan"))
+        system_efficiency = (num_emitted_tokens / max_num_emitted_tokens
+                             if max_num_emitted_tokens > 0 else float("nan"))
+        logger.info(
+            "Speculative metrics: "
+            "Draft acceptance rate: %.3f, "
+            "System efficiency: %.3f, "
+            "Number of speculative tokens: %d, "
+            "Number of accepted tokens: %d, "
+            "Number of draft tokens: %d, "
+            "Number of emitted tokens: %d.", draft_acceptance_rate,
+            system_efficiency, self.num_spec_tokens, num_accepted_tokens,
+            num_draft_tokens, num_emitted_tokens)
+        self.reset()
+
+
+def get_max_num_emitted_tokens(draft_tokens: int, k: int) -> int:
+    """Calculate the number of emitted tokens, assuming all tokens accepted.
+
+    This is equal to the number of sequences that have been speculated on,
+    times (speculation len + 1). The +1 comes from the bonus token.
+    """
+    # Determine the number of sequences that have been speculated on. Since
+    # the batch size can be variable, we divide by k.
+    assert draft_tokens % k == 0
```

Comment:
This code is from V0 but the assumption does not hold in V1, e.g. the ngram proposer says:

> If there are less than k tokens follow the match, we will return the maximum amount of tokens until the end. 

We'll either need to track `num_spec_seqs` so we can do `max_num_emitted_tokens = num_draft_tokens + num_spec_seqs` or just track `max_num_emitted_tokens` as we go

Reply:
Tracking `num_proposals` now